### PR TITLE
feat(HttpClient): add timings to res

### PR DESCRIPTION
**Pull request by hekike**
Measure timings:

- DNS Lookup
- TCP Connection
- TLS Handshake
- First Byte
- Content Transfer
- Total

### Questions

- Should we put it behind a flag?
  - What's the performance impact of calling `process.hrtime()` multiple times, is there any known?

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -142,7 +142,7 @@ function dtrace() {
 
 describe('restify-client tests', function () {
 
-    before(function (callback) {
+    beforeEach(function (callback) {
```
Comment (hekike) [inline test/index.js]: To prevent keep-alive connections

**Comment by DonutEspresso** [comment]
Love this idea! Assuming no significant perf impacts I'm all for this. The only thing I wonder about is if there might be some value in having an API/getter for this vs just plopping it on the req? Otherwise 👍

**Comment by hekike** [comment]
@DonutEspresso I turned it into a getter! ;)

**Comment by DonutEspresso** [comment]
I was looking to see if there were other metrics that might be useful - `retries` may also be useful since the client by default attempts to retry up to 4 times to establish a connection (i.e., it would be good to know on which did this request succeed in establishing a connection?).

I know this metric isn't strictly timing, so we could punt this to another PR.

**Comment by hekike** [comment]
@DonutEspresso awesome, I see you named it to `req.attemtpts()`, and I named it to `res.getTimings()`, I think we should have some alignment about the naming here. I don't feel strongly about the `get` prefix. What do you think?

PS: seems to be that Node.js prefers to use add the get prefix: 
https://nodejs.org/dist/latest-v8.x/docs/api/http.html#http_response_getheader_name

**Comment by DonutEspresso** [comment]
`get` prefix works for me! I'll make that change in the other PR.

I also think we should align on these being on the same object (`req` or `res`?). I'm somewhat partial to `req` since that's how we generally speak about it, e.g., "how long did that request take?"  It also makes sense for some things like `attempts` as well. What do you think?

**Comment by hekike** [comment]
@DonutEspresso I agree, I'm going to move this to `req`!

**Comment by hekike** [comment]
@DonutEspresso rebased and moved to `req`.

File: lib/helpers/timings.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+* Get duration in milliseconds from two process.hrtime()
+* @function getHrTimeDurationInMs
+* @param {Array} startTime - [seconds, nanoseconds]
+* @param {Array} endTime - [seconds, nanoseconds]
+* @returns {Number|null} durationInMs
+*/
+function getHrTimeDurationInMs (startTime, endTime) {
+    if (startTime === null || endTime === null) {
```
Comment (DonutEspresso) [inline lib/helpers/timings.js]: When would these be null? In the case of incomplete/aborted requests that cause only parts of the timing to be recorded?

File: lib/helpers/timings.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+* @param {Number|undefined} eventTimes.tlsHandshakeAt - TLS handshake finished
+* @param {Number} eventTimes.firstByteAt - First byte arrived
+* @param {Number} eventTimes.endAt - Request ended
+* @returns {Object} timings - { dnsLookup, tcpConnection, tlsHandshake,
+*                               firstByte, contentTransfer, total }
+*/
+function getTimings (eventTimes) {
+    return {
+        // There is no DNS lookup with IP address
+        dnsLookup: eventTimes.dnsLookupAt !== null
```
Comment (DonutEspresso) [inline lib/helpers/timings.js]: If `getHrTimeDurationInMs` is returning null already, do you need the additional null check here?

**Comment by DonutEspresso** [comment]
Left some minor comments, otherwise LGTM! ccing other folks for visibility @retrohacker @rajatkumar

File: lib/helpers/timings.js (hunk 2 above)
Comment (hekike) [reply lib/helpers/timings.js]: When for example url is an IP so there is no DNS resolve phase, or no TLS etc.
It's not related to the request start/end time, it's more like a from-to for two htimes, maybe the naming is misleading.

**Comment by DonutEspresso** [comment]
Awesome, this looks great. Let's get this in, we can iterate on it if we need to.

**Pull request merged**